pkg/shell/cockpit-docker.js

```diff
@@ -773,29 +774,251 @@ PageRunImage.prototype = {
         this.containers = null;
     },
 
+    /*
+     * validation functionality for the run image dialog
+     *
+     * warning:
+     *   - an expose port row has only one entry (container or host), the other will be flagged
+     *   - a linked container has no alias or an alias is given for no link
+     *
+     * error:
+     *   - a port is used more than once (same port/protocol exposed on container, same port/protocol used on host)
+     *   - a port number is invalid
+     *   - an alias for a linked container is used more than once
+     *
+     * any errors will result in a disabled 'run' button
+     */
+    update: function(behavior, section) {
```

Comment:
Every time the update function is called all the functions here are redefined. This code is run repeatedly and often.

So it seems a bit extreme given that none of the functions definitions are dependent on either of the arguments to the .update() function. Suggest breaking out the helper functions into the parent scope.